Mathlib/GroupTheory/GroupAction/SubMulAction.lean

```diff
@@ -0,0 +1,385 @@
+  rfl
+#align set_like.mk_smul_mk SetLike.mk_smul_mk
+#align set_like.mk_vadd_mk SetLike.mk_vadd_mk
+
+@[to_additive]
+theorem smul_def (r : R) (x : s) : r • x = ⟨r • x, smul_mem r x.2⟩ :=
+  rfl
+#align set_like.smul_def SetLike.smul_def
+#align set_like.vadd_def SetLike.vadd_def
+
+@[simp]
+theorem forall_smul_mem_iff {R M S : Type _} [Monoid R] [MulAction R M] [SetLike S M]
+    [SMulMemClass S R M] {N : S} {x : M} : (∀ a : R, a • x ∈ N) ↔ x ∈ N :=
+  ⟨fun h => by simpa using h 1, fun h a => SMulMemClass.smul_mem a h⟩
+#align set_like.forall_smul_mem_iff SetLike.forall_smul_mem_iff
+
+end SetLike
+
+/-- A SubMulAction is a set which is closed under scalar multiplication.  -/
+structure SubMulAction (R : Type u) (M : Type v) [SMul R M] : Type v where
+  /-- The underlying set of a `SubMulAction`. -/
+  carrier : Set M
+  /-- The carrier set is closed under scalar multiplication. -/
+  smul_mem' : ∀ (c : R) {x : M}, x ∈ carrier → c • x ∈ carrier
+#align sub_mul_action SubMulAction
+
+namespace SubMulAction
+
+variable [SMul R M]
+
+instance : SetLike (SubMulAction R M) M :=
+  ⟨SubMulAction.carrier, fun p q h => by cases p; cases q; congr ⟩
+
+instance : SMulMemClass (SubMulAction R M) R M where smul_mem := smul_mem' _
+
+@[simp]
+theorem mem_carrier {p : SubMulAction R M} {x : M} : x ∈ p.carrier ↔ x ∈ (p : Set M) :=
+  Iff.rfl
+#align sub_mul_action.mem_carrier SubMulAction.mem_carrier
+
+@[ext]
+theorem ext {p q : SubMulAction R M} (h : ∀ x, x ∈ p ↔ x ∈ q) : p = q :=
+  SetLike.ext h
+#align sub_mul_action.ext SubMulAction.ext
+
+/-- Copy of a sub_mul_action with a new `carrier` equal to the old one. Useful to fix definitional
+equalities.-/
+protected def copy (p : SubMulAction R M) (s : Set M) (hs : s = ↑p) : SubMulAction R M
+    where
+  carrier := s
+  smul_mem' := hs.symm ▸ p.smul_mem'
+#align sub_mul_action.copy SubMulAction.copy
+
+@[simp]
+theorem coe_copy (p : SubMulAction R M) (s : Set M) (hs : s = ↑p) : (p.copy s hs : Set M) = s :=
+  rfl
+#align sub_mul_action.coe_copy SubMulAction.coe_copy
+
+theorem copy_eq (p : SubMulAction R M) (s : Set M) (hs : s = ↑p) : p.copy s hs = p :=
+  SetLike.coe_injective hs
+#align sub_mul_action.copy_eq SubMulAction.copy_eq
+
+instance : Bot (SubMulAction R M) where
+  bot :=
+    { carrier := ∅
+      smul_mem' := fun _c h => Set.not_mem_empty h }
+
+instance : Inhabited (SubMulAction R M) :=
+  ⟨⊥⟩
+
+end SubMulAction
+
+namespace SubMulAction
+
+section SMul
+
+variable [SMul R M]
+
+variable (p : SubMulAction R M)
+
+variable {r : R} {x : M}
+
+theorem smul_mem (r : R) (h : x ∈ p) : r • x ∈ p :=
+  p.smul_mem' r h
+#align sub_mul_action.smul_mem SubMulAction.smul_mem
+
+instance : SMul R p where smul c x := ⟨c • x.1, smul_mem _ c x.2⟩
+
+variable {p}
+
+@[simp, norm_cast]
+theorem val_smul (r : R) (x : p) : (↑(r • x) : M) = r • (x : M) :=
+  rfl
+#align sub_mul_action.coe_smul SubMulAction.val_smul
+
+-- porting note: no longer needed because of defeq structure eta
```

Comment:
I'm not sure why the linter doesn't like this lemma. From where I'm sitting the head symbol is not x. Also, I don't think this has anything to do with structural eta

Reply:
Linter says
```lean
/- The `simpVarHead` linter reports:
LEFT-HAND SIDE HAS VARIABLE AS HEAD SYMBOL.
Some simp lemmas have a variable as head symbol of the left-hand side (after whnfR): -/
#check @coe_mk /- Left-hand side has variable as head symbol: x -/
```
probably because
`Subtype.val { val := x, property := hx } = x`

Reply:
Yes, my understanding was that this is exactly due to the structural eta defeq `Subtype.val { val := x, property := hx } = x`. It was my understanding that we have been removing these `coe_mk` lemmas across the library. If you search the docs for `coe_mk`, you'll find a bunch, but I think these are for homs and set-likes, where the `coe` is `FunLike.coe` or `SetLike.coe` and hence the structural eta doesn't apply.